File: solutions/java/src/main/java/com/epi/StackSorting.java
package com.epi;

import java.util.LinkedList;
import java.util.Random;

public class StackSorting {
  // @include
  public static void sort(LinkedList<Integer> S) {
    if (!S.isEmpty()) {
      Integer val = S.pop();
      sort(S);
      insert(val, S);
    }
  }

  private static void insert(Integer val, LinkedList<Integer> S) {
    if (S.isEmpty() || S.peek().compareTo(val) <= 0) {
      S.push(val);
    } else {
      Integer f = S.pop();
      insert(val, S);
      S.push(f);
    }
  }
  // @exclude

  private static void simpleTest() {
    LinkedList<Integer> S = new LinkedList<>();
    S.push(1);
    sort(S);
    assert(S.peek() == 1);
    S.push(0);
    sort(S);
    assert(S.peek() == 1);
    S.pop();
    assert(S.peek() == 0);
    S.pop();
    assert(S.isEmpty());
    S.push(-1);
    S.push(1);
    S.push(0);
    sort(S);
    assert(S.peek() == 1);
    S.pop();
    assert(S.peek() == 0);
    S.pop();
    assert(S.peek() == -1);
    S.pop();
    assert(S.isEmpty());
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = r.nextInt(10000) + 1;
      }

      LinkedList<Integer> S = new LinkedList<>();
      for (int i = 0; i < n; i++) {
        S.push(r.nextInt(1000000));
      }
      sort(S);
      int pre = Integer.MAX_VALUE;
      while (!S.isEmpty()) {
        assert(pre >= S.peek());
        System.out.println(S.peek());
        pre = S.pop();
      }
    }
  }
}

File: static/python/stack_sorting.py
# Stack_sorting.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import random


# @include
def sort(S):
    if S:
        e = S.pop()
        sort(S)
        insert(e, S)


def insert(e, S):
    if not S or S[-1] <= e:
        S.append(e)
    else:
        f = S.pop()
        insert(e, S)
        S.append(f)


# @exclude


def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
        else:
            n = random.randint(1, sys.getrecursionlimit() - 3)
        S = [random.randint(0, 999999) for i in range(n)]
        sort(S)
        pre = float('inf')
        while S:
            assert pre >= S[-1]
            print(S[-1])
            pre = S.pop()


if __name__ == '__main__':
    main()
